### PR TITLE
[SYCL] Ignore 'used' attribute when emitting sycl device code

### PR DESCRIPTION
**Pull request by premanandrao**
Signed-off-by: Premanand M Rao <premanand.m.rao@intel.com>

### CODE REVIEW

File: clang/lib/AST/ASTContext.cpp [hunk 1]
```diff
@@ -10446,14 +10446,18 @@ bool ASTContext::DeclMustBeEmitted(const Decl *D) {
   if (D->hasAttr<WeakRefAttr>())
     return false;
 
-  // Aliases and used decls are required.
-  if (D->hasAttr<AliasAttr>() || D->hasAttr<UsedAttr>())
+  // Aliases are required.
+  if (D->hasAttr<AliasAttr>())
     return true;
```
Comment (Fznamznon) [inline clang/lib/AST/ASTContext.cpp:10451]: I don't think that we should split them up.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 2]
```diff
@@ -3865,13 +3865,13 @@ LangAS CodeGenModule::GetGlobalVarAddressSpace(const VarDecl *D) {
   }
 
   if (LangOpts.SYCLIsDevice) {
+    if (!D)
+      return LangAS::opencl_global;
```
Comment (Fznamznon) [inline clang/lib/CodeGen/CodeGenModule.cpp:3869]: I understand that this is a useful bugfix but it seems to be a bit unrelated to this patch. Can we do it separately?

File: clang/test/CodeGenSYCL/addrspace_ctor_dtor.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -disable-llvm-passes %s -emit-llvm -o -  | FileCheck %s
```
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/addrspace_ctor_dtor.cpp:1]: Emitting llvm IR without passing spir-based triple will result in llvm_unreachable somewhere in CodeGen, AFAIK.
BTW, think we need to rename test, so the name will reflect to changes related to used attribute.

File: clang/lib/AST/ASTContext.cpp (hunk 1 above)
Comment (premanandrao) [reply clang/lib/AST/ASTContext.cpp]: Okay, I have recombined them and moved that after the SYCL check.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 2 above)
Comment (premanandrao) [reply clang/lib/CodeGen/CodeGenModule.cpp]: That makes sense, I will do that.

File: clang/test/CodeGenSYCL/addrspace_ctor_dtor.cpp (hunk 3 above)
Comment (premanandrao) [reply clang/test/CodeGenSYCL/addrspace_ctor_dtor.cpp]: Done.

**Comment by bader** [comment]
@premanandrao, what is the motivation for this change?
At some point I was thinking that we might use this attribute to implement `SYCL_EXTERNAL` macro to avoid adding another custom attribute.
Is it used in some legacy code, which we compile "as is" in SYCL mode?

**Comment by Fznamznon** [comment]
> @premanandrao, what is the motivation for this change?
> At some point I was thinking that we might use this attribute to implement `SYCL_EXTERNAL` macro to avoid adding another custom attribute.
> Is it used in some legacy code, which we compile "as is" in SYCL mode?

The original problem was with crash due to address space mismatch. But there was emission of some global variable with initializers forced by `used` attribute, I don't think that `used` attribute should force emission of device code unless SYCL spec requires it (and it doesn't).
I also don't think that it is a good idea to use `used` attrubute to implement `SYCL_EXTERNAL` because they may have different semantics.

**Comment by erichkeane** [comment]
> 
> 
> > @premanandrao, what is the motivation for this change?
> > At some point I was thinking that we might use this attribute to implement `SYCL_EXTERNAL` macro to avoid adding another custom attribute.
> > Is it used in some legacy code, which we compile "as is" in SYCL mode?
> 
> The original problem was with crash due to address space mismatch. But there was emission of some global variable with initializers forced by `used` attribute, I don't think that `used` attribute should force emission of device code unless SYCL spec requires it (and it doesn't).
> I also don't think that it is a good idea to use `used` attrubute to implement `SYCL_EXTERNAL` because they may have different semantics.

I agree that we shouldn't use 'used' as SYCL_EXTERNAL.  It has the problem of being used in existing code, so we'll end up having some odd variety of non-SYCL_EXTERNAL code be valid/linkable/callable, simply because the user did attribute-used in the past.

I believe a new/novel attribute is the correct way to go here.

**Comment by premanandrao** [comment]
> I believe a new/novel attribute is the correct way to go here.

+1

**Comment by premanandrao** [comment]
Thanks @bader
